atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStoreTest.java

```diff
@@ -0,0 +1,111 @@
+    private static final long WALL_CLOCK_2 = 200L;
+    private static final long WALL_CLOCK_3 = 300L;
+    private static final long WALL_CLOCK_BETWEEN_1_AND_2 = 150L;
+    private static final long WALL_CLOCK_BETWEEN_2_AND_3 = 250L;
+
+    private static final ImmutableMap<Long, Long> PUNCHER_HISTORY = ImmutableMap.of(
+            TIMESTAMP_1, WALL_CLOCK_1,
+            TIMESTAMP_2, WALL_CLOCK_2,
+            TIMESTAMP_3, WALL_CLOCK_3);
+
+    private final PuncherStore puncherStore = KeyValueServicePuncherStore.create(new InMemoryKeyValueService(false));
+
+    @Before
+    public void setUp() {
+        PUNCHER_HISTORY.entrySet().forEach(entry -> puncherStore.put(entry.getKey(), entry.getValue()));
+    }
+
+    @Test
+    public void getTimestampForMillisReturnsMinValueIfQueryingClockTimeBeforeFirstPunch() {
+        assertThat(puncherStore.get(WALL_CLOCK_1 - 1)).isEqualTo(Long.MIN_VALUE);
+        assertThat(puncherStore.get(0L)).isEqualTo(Long.MIN_VALUE);
+    }
+
+    @Test
+    public void getTimestampForMillisReturnsTimestampIfQueryingPreciseClockTime() {
+        PUNCHER_HISTORY.entrySet().forEach(
+                entry -> assertThat(puncherStore.get(entry.getValue())).isEqualTo(entry.getKey()));
+    }
+
+    @Test
+    public void getTimestampForMillisReturnsLastTimestampKnownToOccurBeforeQueriedTime() {
+        PUNCHER_HISTORY.entrySet().forEach(
+                entry -> assertThat(puncherStore.get(entry.getValue() + 1)).isEqualTo(entry.getKey()));
+    }
+
+    @Test
+    public void getMillisForTimestampReturnsZeroIfQueryingTimestampBeforeFirstPunch() {
+        assertThat(puncherStore.getMillisForTimestamp(Long.MIN_VALUE)).isEqualTo(0L);
+        assertThat(puncherStore.getMillisForTimestamp(TIMESTAMP_1 - 1)).isEqualTo(0L);
+    }
+
+    @Test
+    public void getMillisForTimestampReturnsClockTimeIfQueryingPreciseTimestamp() {
+        PUNCHER_HISTORY.entrySet().forEach(
+                entry -> assertThat(puncherStore.getMillisForTimestamp(entry.getKey())).isEqualTo(entry.getValue()));
+    }
+
+    @Test
+    public void getMillisForTimestampReturnsLastTimeKnownToOccurBeforeQueriedTimestamp() {
+        PUNCHER_HISTORY.entrySet().forEach(
+                entry -> assertThat(puncherStore.getMillisForTimestamp(entry.getKey() + 1))
+                        .isEqualTo(entry.getValue()));
+    }
+
+    @Test
+    public void handlesPunchesNotInSequentialOrder() {
+        assertThat(puncherStore.getMillisForTimestamp(TIMESTAMP_BETWEEN_1_AND_2))
+                .isEqualTo(WALL_CLOCK_1);
+        puncherStore.put(TIMESTAMP_BETWEEN_1_AND_2, WALL_CLOCK_BETWEEN_1_AND_2);
+        assertThat(puncherStore.getMillisForTimestamp(TIMESTAMP_BETWEEN_1_AND_2))
+                .isEqualTo(WALL_CLOCK_BETWEEN_1_AND_2);
+    }
+
+    @Test
+    public void returnsGreatestPunchedTimeBeforeTimestampEvenIfNotAssociatedWithGreatestEligibleTimestamp() {
+        puncherStore.put(TIMESTAMP_BETWEEN_1_AND_2, WALL_CLOCK_BETWEEN_2_AND_3);
```

Comment:
Yep, agree that having a separate map would be much more explicit. I'll change this to do that instead.